File: lib/src/main/java/net/mm2d/upnp/HttpMessageDelegate.java
/*
 * Copyright (c) 2018 大前良介 (OHMAE Ryosuke)
 *
 * This software is released under the MIT License.
 * http://opensource.org/licenses/MIT
 */

package net.mm2d.upnp;

import net.mm2d.log.Log;
import net.mm2d.util.TextUtils;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.io.UnsupportedEncodingException;
import java.util.Arrays;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;

/**
 * HttpMessageの共通実装。
 *
 * @author <a href="mailto:ryo@mm2d.net">大前良介 (OHMAE Ryosuke)</a>
 */
class HttpMessageDelegate implements HttpMessage {
    private static final int BUFFER_SIZE = 1500;
    private static final int DEFAULT_CHUNK_SIZE = 1024;
    private static final int CR = 0x0d;
    private static final int LF = 0x0a;
    private static final byte[] CRLF = new byte[]{(byte) CR, (byte) LF};
    private static final String EOL = "\r\n";
    private static final String CHARSET = "utf-8";

    interface StartLineProcessor {
        void setStartLine(@Nonnull String line);

        @Nonnull
        String getStartLine();
    }

    @Nonnull
    private final HttpHeaders mHeaders;
    @Nonnull
    private String mVersion = Http.DEFAULT_HTTP_VERSION;
    @Nullable
    private byte[] mBodyBinary;
    @Nullable
    private String mBody;
    @Nonnull
    private final StartLineProcessor mStartLineProcessor;

    public HttpMessageDelegate(@Nonnull final StartLineProcessor processor) {
        mStartLineProcessor = processor;
        mHeaders = new HttpHeaders();
    }

    public HttpMessageDelegate(
            @Nonnull final StartLineProcessor processor,
            @Nonnull final HttpMessageDelegate original) {
        mStartLineProcessor = processor;
        mHeaders = new HttpHeaders(original.mHeaders);
        mBodyBinary = original.mBodyBinary == null ? null :
                Arrays.copyOf(original.mBodyBinary, original.mBodyBinary.length);
        mBody = original.mBody;
    }

    @Nullable
    @Override
    public String getStartLine() {
        return mStartLineProcessor.getStartLine();
    }

    @Nonnull
    @Override
    public HttpMessage setStartLine(@Nonnull final String line) {
        mStartLineProcessor.setStartLine(line);
        return this;
    }

    @Nonnull
    @Override
    public String getVersion() {
        return mVersion;
    }

    @Nonnull
    @Override
    public HttpMessage setVersion(@Nonnull final String version) {
        mVersion = version;
        return this;
    }

    @Nonnull
    @Override
    public HttpMessage setHeader(
            @Nonnull final String name,
            @Nonnull final String value) {
        mHeaders.put(name, value);
        return this;
    }

    @Nonnull
    @Override
    public HttpMessage setHeaderLine(@Nonnull final String line) {
        final int pos = line.indexOf(':');
        if (pos < 0) {
            return this;
        }
        final String name = line.substring(0, pos).trim();
        final String value = line.substring(pos + 1).trim();
        return setHeader(name, value);
    }

    @Nullable
    @Override
    public String getHeader(@Nonnull final String name) {
        return mHeaders.get(name);
    }

    @Override
    public boolean isChunked() {
        return mHeaders.containsValue(Http.TRANSFER_ENCODING, Http.CHUNKED);
    }

    @Override
    public boolean isKeepAlive() {
        if (mVersion.equals(Http.HTTP_1_0)) {
            return mHeaders.containsValue(Http.CONNECTION, Http.KEEP_ALIVE);
        }
        return !mHeaders.containsValue(Http.CONNECTION, Http.CLOSE);
    }

    @Override
    public int getContentLength() {
        final String len = mHeaders.get(Http.CONTENT_LENGTH);
        if (len != null) {
            try {
                return Integer.parseInt(len);
            } catch (final NumberFormatException e) {
                Log.w(e);
            }
        }
        return 0;
    }

    @Nonnull
    @Override
    public HttpMessage setBody(@Nullable final String body) {
        return setBody(body, false);
    }

    @Nonnull
    @Override
    public HttpMessage setBody(
            @Nullable final String body,
            final boolean withContentLength) {
        return setBodyInner(body, null, withContentLength);
    }

    @Nonnull
    @Override
    public HttpMessage setBodyBinary(@Nullable final byte[] body) {
        return setBodyBinary(body, false);
    }

    @Nonnull
    @Override
    public HttpMessage setBodyBinary(
            @Nullable final byte[] body,
            final boolean withContentLength) {
        return setBodyInner(null, body, withContentLength);
    }

    @Nonnull
    private HttpMessage setBodyInner(
            @Nullable final String string,
            @Nullable final byte[] binary,
            final boolean withContentLength) {
        mBody = string;
        if (string == null) {
            mBodyBinary = binary;
        } else if (string.length() == 0) {
            mBodyBinary = new byte[0];
        } else {
            try {
                mBodyBinary = getBytes(string);
            } catch (final UnsupportedEncodingException e) {
                Log.w(e);
            }
        }
        if (withContentLength) {
            final int length = mBodyBinary == null ? 0 : mBodyBinary.length;
            setHeader(Http.CONTENT_LENGTH, String.valueOf(length));
        }
        return this;
    }

    // VisibleForTesting
    @Nonnull
    byte[] getBytes(@Nonnull final String string) throws UnsupportedEncodingException {
        return string.getBytes(CHARSET);
    }

    @Nullable
    @Override
    public String getBody() {
        if (mBody == null && mBodyBinary != null) {
            mBody = decode(mBodyBinary);
        }
        return mBody;
    }

    @Nullable
    private String decode(@Nonnull final byte[] binary) {
        if (binary.length == 0) {
            return "";
        }
        try {
            return newString(binary);
        } catch (final Exception e) {
            // for bug in Android Sdk, ArrayIndexOutOfBoundsException may occur.
            Log.w(e);
        }
        return null;
    }

    // VisibleForTesting
    @Nonnull
    String newString(@Nonnull final byte[] binary) throws UnsupportedEncodingException {
        return new String(binary, CHARSET);
    }

    @Nullable
    @Override
    public byte[] getBodyBinary() {
        return mBodyBinary;
    }

    @Nonnull
    private String getHeaderString() {
        final StringBuilder sb = new StringBuilder();
        sb.append(getStartLine());
        sb.append(EOL);
        for (final HttpHeaders.Entry entry : mHeaders.values()) {
            sb.append(entry.getName());
            sb.append(": ");
            sb.append(entry.getValue());
            sb.append(EOL);
        }
        sb.append(EOL);
        return sb.toString();
    }

    // VisibleForTesting
    @Nonnull
    byte[] getHeaderBytes() {
        try {
            return getBytes(getHeaderString());
        } catch (final UnsupportedEncodingException e) {
            Log.w(e);
        }
        return new byte[0];
    }

    @Nonnull
    @Override
    public String getMessageString() {
        final StringBuilder sb = new StringBuilder();
        sb.append(getStartLine());
        sb.append(EOL);
        for (final HttpHeaders.Entry entry : mHeaders.values()) {
            sb.append(entry.getName());
            sb.append(": ");
            sb.append(entry.getValue());
            sb.append(EOL);
        }
        sb.append(EOL);
        final String body = getBody();
        if (!TextUtils.isEmpty(body)) {
            sb.append(body);
        }
        return sb.toString();
    }

    @Override
    public void writeData(@Nonnull final OutputStream outputStream) throws IOException {
        outputStream.write(getHeaderBytes());
        if (mBodyBinary != null) {
            if (isChunked()) {
                writeChunkedBody(outputStream, mBodyBinary);
            } else {
                outputStream.write(mBodyBinary);
            }
        }
        outputStream.flush();
    }

    private void writeChunkedBody(
            @Nonnull final OutputStream os,
            @Nonnull final byte[] binary)
            throws IOException {
        int offset = 0;
        while (offset < binary.length) {
            final int size = Math.min(DEFAULT_CHUNK_SIZE, binary.length - offset);
            writeChunkSize(os, size);
            os.write(binary, offset, size);
            os.write(CRLF);
            offset += size;
        }
        writeChunkSize(os, 0);
        os.write(CRLF);
    }

    private void writeChunkSize(
            @Nonnull final OutputStream os,
            final int size) throws IOException {
        os.write(getBytes(Integer.toHexString(size)));
        os.write(CRLF);
    }

    @Override
    public HttpMessage readData(@Nonnull final InputStream inputStream) throws IOException {
        readStartLine(inputStream);
        readHeaders(inputStream);
        if (isChunked()) {
            readChunkedBody(inputStream);
        } else {
            readBody(inputStream);
        }
        return this;
    }

    private void readStartLine(@Nonnull final InputStream is) throws IOException {
        final String startLine = readLine(is);
        if (TextUtils.isEmpty(startLine)) {
            throw new IOException("Illegal start line:" + startLine);
        }
        try {
            setStartLine(startLine);
        } catch (final IllegalArgumentException e) {
            throw new IOException("Illegal start line:" + startLine);
        }
    }

    private void readHeaders(@Nonnull final InputStream is) throws IOException {
        while (true) {
            final String line = readLine(is);
            if (line.isEmpty()) {
                break;
            }
            setHeaderLine(line);
        }
    }

    private void readBody(@Nonnull final InputStream is) throws IOException {
        final ByteArrayOutputStream baos = new ByteArrayOutputStream();
        int length = getContentLength();
        final byte[] buffer = new byte[BUFFER_SIZE];
        while (length > 0) {
            final int stroke = length > buffer.length ? buffer.length : length;
            final int size = is.read(buffer, 0, stroke);
            if (size < 0) {
                throw new IOException("can't read from InputStream");
            }
            baos.write(buffer, 0, size);
            length -= size;
        }
        setBodyBinary(baos.toByteArray());
    }

    private void readChunkedBody(@Nonnull final InputStream is) throws IOException {
        final ByteArrayOutputStream baos = new ByteArrayOutputStream();
        while (true) {
            int length = readChunkSize(is);
            if (length == 0) {
                readLine(is);
                break;
            }
            final byte[] buffer = new byte[BUFFER_SIZE];
            while (length > 0) {
                final int stroke = length > buffer.length ? buffer.length : length;
                final int size = is.read(buffer, 0, stroke);
                if (size < 0) {
                    throw new IOException("can't read from InputStream");
                }
                baos.write(buffer, 0, size);
                length -= size;
            }
            readLine(is);
        }
        setBodyBinary(baos.toByteArray());
    }

    private int readChunkSize(@Nonnull final InputStream is) throws IOException {
        final String line = readLine(is);
        if (TextUtils.isEmpty(line)) {
            throw new IOException("Can not read chunk size!");
        }
        final String chunkSize = line.split(";", 2)[0];
        try {
            return Integer.parseInt(chunkSize, 16);
        } catch (final NumberFormatException e) {
            throw new IOException("Chunk format error!", e);
        }
    }

    @Nonnull
    private static String readLine(@Nonnull final InputStream is) throws IOException {
        final ByteArrayOutputStream baos = new ByteArrayOutputStream();
        while (true) {
            final int b = is.read();
            if (b < 0) {
                if (baos.size() == 0) {
                    throw new IOException("can't read from InputStream");
                }
                break;
            }
            if (b == LF) {
                break;
            }
            if (b == CR) {
                continue;
            }
            baos.write(b);
        }
        return baos.toString(CHARSET);
    }

    @Override
    @Nonnull
    public String toString() {
        return getMessageString();
    }
}

File: mmupnp/src/main/java/net/mm2d/upnp/internal/message/HttpMessageDelegate.kt
/*
 * Copyright (c) 2019 大前良介 (OHMAE Ryosuke)
 *
 * This software is released under the MIT License.
 * http://opensource.org/licenses/MIT
 */

package net.mm2d.upnp.internal.message

import net.mm2d.log.Logger
import net.mm2d.upnp.Http
import net.mm2d.upnp.HttpMessage
import java.io.*
import kotlin.math.min

/**
 * Common implementation of [HttpMessage].
 *
 * @author [大前良介 (OHMAE Ryosuke)](mailto:ryo@mm2d.net)
 */
internal class HttpMessageDelegate(
    private val startLineDelegate: StartLineDelegate,
    original: HttpMessageDelegate? = null
) : HttpMessage {
    internal interface StartLineDelegate {
        var version: String

        fun getStartLine(): String

        fun setStartLine(startLine: String)
    }

    private val headers: HttpHeaders
    private var body: String? = null
    private var bodyBinary: ByteArray? = null

    init {
        if (original == null) {
            headers = HttpHeaders()
        } else {
            headers = HttpHeaders(original.headers)
            body = original.body
            bodyBinary = original.bodyBinary?.copyOf()
        }
    }

    override val startLine: String?
        get() = startLineDelegate.getStartLine()
    override val version: String
        get() = startLineDelegate.version
    override val isChunked: Boolean
        get() = headers.containsValue(Http.TRANSFER_ENCODING, Http.CHUNKED)
    override val contentLength: Int
        get() = headers[Http.CONTENT_LENGTH]?.toIntOrNull() ?: 0

    // VisibleForTesting
    @Throws(UnsupportedEncodingException::class)
    fun ByteArray.newString(): String {
        return String(this, CHARSET)
    }

    private fun getHeaderString(): String {
        return getHeaderStringBuilder().toString()
    }

    private fun getHeaderStringBuilder(): StringBuilder {
        return StringBuilder().also { sb ->
            sb.append(startLine)
            sb.append(EOL)
            headers.values().forEach {
                sb.append(it.name)
                sb.append(": ")
                sb.append(it.value)
                sb.append(EOL)
            }
            sb.append(EOL)
        }
    }

    // VisibleForTesting
    internal fun getHeaderBytes(): ByteArray {
        try {
            return getBytes(getHeaderString())
        } catch (e: UnsupportedEncodingException) {
            Logger.w(e)
        }
        return ByteArray(0)
    }

    override fun isKeepAlive(): Boolean = if (version == Http.HTTP_1_0) {
        headers.containsValue(Http.CONNECTION, Http.KEEP_ALIVE)
    } else {
        !headers.containsValue(Http.CONNECTION, Http.CLOSE)
    }

    override fun setStartLine(line: String) {
        startLineDelegate.setStartLine(line)
    }

    override fun setVersion(version: String) {
        startLineDelegate.version = version
    }

    override fun setHeader(name: String, value: String) {
        headers.put(name, value)
    }

    override fun setHeaderLine(line: String) {
        val section = line.split(":", limit = 2)
        if (section.size < 2) {
            return
        }
        return setHeader(section[0].trim(), section[1].trim())
    }

    override fun getHeader(name: String): String? {
        return headers[name]
    }

    override fun getBody(): String? {
        return body ?: bodyBinary?.decode()?.also { body = it }
    }

    private fun ByteArray.decode(): String? {
        if (isEmpty()) return ""
        try {
            return newString()
        } catch (e: Exception) {
            // for bug in Android Sdk, ArrayIndexOutOfBoundsException may occur.
            Logger.w(e)
        }
        return null
    }

    override fun setBody(body: String?, withContentLength: Boolean) {
        setBodyInner(body, null, withContentLength)
    }

    override fun getBodyBinary(): ByteArray? {
        return bodyBinary
    }

    override fun setBodyBinary(body: ByteArray?, withContentLength: Boolean) {
        setBodyInner(null, body, withContentLength)
    }

    private fun setBodyInner(string: String?, binary: ByteArray?, withContentLength: Boolean) {
        body = string
        when {
            string == null ->
                bodyBinary = binary
            string.isEmpty() ->
                bodyBinary = ByteArray(0)
            else -> try {
                bodyBinary = getBytes(string)
            } catch (e: UnsupportedEncodingException) {
                Logger.w(e)
            }
        }
        if (withContentLength) {
            setHeader(Http.CONTENT_LENGTH, (bodyBinary?.size ?: 0).toString())
        }
    }

    // VisibleForTesting
    @Throws(UnsupportedEncodingException::class)
    internal fun getBytes(string: String): ByteArray {
        return string.toByteArray(CHARSET)
    }

    override fun getMessageString(): String {
        val body = body
        return getHeaderStringBuilder().also {
            if (!body.isNullOrEmpty()) {
                it.append(body)
            }
        }.toString()
    }

    @Throws(IOException::class)
    override fun writeData(outputStream: OutputStream) {
        outputStream.write(getHeaderBytes())
        bodyBinary?.let {
            if (isChunked) {
                writeChunkedBody(outputStream, it)
            } else {
                outputStream.write(it)
            }
        }
        outputStream.flush()
    }

    @Throws(IOException::class)
    private fun writeChunkedBody(outputStream: OutputStream, binary: ByteArray) {
        var offset = 0
        while (offset < binary.size) {
            val size = min(DEFAULT_CHUNK_SIZE, binary.size - offset)
            writeChunkSize(outputStream, size)
            outputStream.write(binary, offset, size)
            outputStream.write(CRLF)
            offset += size
        }
        writeChunkSize(outputStream, 0)
        outputStream.write(CRLF)
    }

    @Throws(IOException::class)
    private fun writeChunkSize(outputStream: OutputStream, size: Int) {
        outputStream.write(getBytes(Integer.toHexString(size)))
        outputStream.write(CRLF)
    }

    @Throws(IOException::class)
    override fun readData(inputStream: InputStream) {
        readStartLine(inputStream)
        readHeaders(inputStream)
        if (isChunked) {
            readChunkedBody(inputStream)
        } else {
            readBody(inputStream)
        }
    }

    @Throws(IOException::class)
    private fun readStartLine(inputStream: InputStream) {
        val startLine = readLine(inputStream)
        if (startLine.isEmpty()) {
            throw IOException("Illegal start line:$startLine")
        }
        try {
            setStartLine(startLine)
        } catch (e: IllegalArgumentException) {
            throw IOException("Illegal start line:$startLine")
        }
    }

    @Throws(IOException::class)
    private fun readHeaders(inputStream: InputStream) {
        while (true) {
            val line = readLine(inputStream)
            if (line.isEmpty()) {
                break
            }
            setHeaderLine(line)
        }
    }

    @Throws(IOException::class)
    private fun readBody(inputStream: InputStream) {
        bodyBinary = ByteArrayOutputStream().also {
            readInputStream(inputStream, it, contentLength)
        }.toByteArray()
    }

    @Throws(IOException::class)
    private fun readChunkedBody(inputStream: InputStream) {
        bodyBinary = ByteArrayOutputStream().also {
            while (true) {
                val length = readChunkSize(inputStream)
                if (length == 0) {
                    readLine(inputStream)
                    break
                }
                readInputStream(inputStream, it, length)
                readLine(inputStream)
            }
        }.toByteArray()
    }

    @Throws(IOException::class)
    private fun readInputStream(inputStream: InputStream, outputStream: OutputStream, length: Int) {
        val buffer = ByteArray(BUFFER_SIZE)
        var remain = length
        while (remain > 0) {
            val stroke = if (remain > buffer.size) buffer.size else remain
            val size = inputStream.read(buffer, 0, stroke)
            if (size < 0) {
                throw IOException("can't read from InputStream")
            }
            outputStream.write(buffer, 0, size)
            remain -= size
        }
    }

    @Throws(IOException::class)
    private fun readChunkSize(inputStream: InputStream): Int {
        val line = readLine(inputStream)
        if (line.isEmpty()) {
            throw IOException("Can not read chunk size!")
        }
        val chunkSize = line.split(";", limit = 2)[0]
        return chunkSize.toIntOrNull(16) ?: throw IOException("Chunk format error! $chunkSize")
    }

    override fun toString(): String {
        return getMessageString()
    }

    companion object {
        private const val BUFFER_SIZE = 1500
        private const val DEFAULT_CHUNK_SIZE = 1024
        private const val CR: Int = '\r'.toInt()
        private const val LF: Int = '\n'.toInt()
        private const val EOL: String = "\r\n"
        private val CRLF = byteArrayOf(CR.toByte(), LF.toByte())
        private val CHARSET = Charsets.UTF_8

        @Throws(IOException::class)
        private fun readLine(inputStream: InputStream): String {
            return ByteArrayOutputStream().also {
                while (true) {
                    val b = inputStream.read()
                    if (b < 0) {
                        if (it.size() == 0) throw IOException("can't read from InputStream")
                        break
                    }
                    if (b == LF) break
                    if (b == CR) continue
                    it.write(b)
                }
            }.toString(CHARSET.name())
        }
    }
}
